# Bioinfomatics Assignments, year 2022/2023
Codes and details. 
## Assignment 1: Sequence Alignment
Implementing the Needleman-Wunsch algorithm, dynamic programming approach for the global sequence alignment, and the Smith-Waterman algorithm, dynamic programming approach for the local sequence alignment, both including the traceback.
- Input: two sequences
- Output: their alignment 
## Assignment 2: Genome Assembly
Finding the shortest common superstring, brute-force and greedy approaches for the suffix-prefix overlap.
- Input: a set of strings
- Output: their shortest common superstring 
## Assignment 3: Read mapping
Implementing the Burrow-Wheeler transform and string matching.
- Input: a genome sequence T and a query P
- Output: all matches of P in T
## Assignment 4: Genome Annotation
Computing the inside and outside CpG island models, using them to score a query sequence.
- Input: CpG sequences, non-CpG sequences, a query string X
- Output: the log-ratio score of X given the inside and outside models
The FASTA file of chromosome 22 is required for the codes to work, which should be stored in the same folder of the python codes as 'chr22.fa' ; it can be downloaded with the following link: [http://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/chr22.fa.gz]
